[Bug]: Mapping binding pattern fails for anonymous record type

### Description

Consider the following code
```ballerina
import ballerina/io;

type Representation record {
    string __typename;
};

public function main() returns error? {
    Representation representation = {__typename: "Product", "variation": {id: "10"}};
    record {
        string id?;
        string sku?;
        string package?;
        record {string id;} variation?;
    } {id, sku, package, variation} = check representation.ensureType(); // this crashes

    io:println(id, variation?.id);
}
```
The above code crashes with following error
```
[2023-02-16 12:59:49,370] SEVERE {b7a.log.crash} - $typedesc$$anonType$_0 
java.lang.NoClassDefFoundError: $typedesc$$anonType$_0
	at test.main(test.bal:0)
	at $_init.$lambda$main$(.)
	at io.ballerina.runtime.internal.scheduling.SchedulerItem.execute(SchedulerItem.java:54)
	at io.ballerina.runtime.internal.scheduling.Scheduler.run(Scheduler.java:324)
	at io.ballerina.runtime.internal.scheduling.Scheduler.runSafely(Scheduler.java:291)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.ClassNotFoundException: $typedesc$$anonType$_0
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	... 6 more
```

### Steps to Reproduce

_No response_

### Affected Version(s)

2201.4.0 RC2, 2201.3.0

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

--- a/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java
+++ b/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java
@@ -1560,6 +1560,7 @@ public class Desugar extends BLangNodeVisitor {
 
     @Override
     public void visit(BLangRecordVariable varNode) {
+        varNode.typeNode = rewrite(varNode.typeNode, env);
         final BLangBlockStmt blockStmt = ASTBuilderUtil.createBlockStmt(varNode.pos);
         String name = anonModelHelper.getNextRecordVarKey(env.enclPkg.packageID);
         final BLangSimpleVariable mapVariable =

--- a/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/variabledef/RecordVariableDefinitionTest.java
+++ b/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/variabledef/RecordVariableDefinitionTest.java
@@ -430,6 +430,11 @@ public class RecordVariableDefinitionTest {
         BRunUtil.invoke(result, "testReadOnlyRecordWithMappingBindingPatternInVarDecl");
     }
 
+    @Test
+    public void testRecordVariableWithAnonymousType() {
+        BRunUtil.invoke(result, "testRecordVariableWithAnonymousType");
+    }
+
     @AfterClass
     public void tearDown() {
         result = null;

--- a/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/variabledef/RecordVariableDefinitionTest.java
+++ b/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/variabledef/RecordVariableDefinitionTest.java
@@ -435,6 +435,11 @@ public class RecordVariableDefinitionTest {
         BRunUtil.invoke(result, "testRecordVariableWithAnonymousType");
     }
 
+    @Test
+    public void testNestedRecordVariableWithAnonymousType() {
+        BRunUtil.invoke(result, "testNestedRecordVariableWithAnonymousType");
+    }
+
     @AfterClass
     public void tearDown() {
         result = null;

--- a/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/variabledef/TupleVariableDefinitionTest.java
+++ b/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/variabledef/TupleVariableDefinitionTest.java
@@ -383,6 +383,11 @@ public class TupleVariableDefinitionTest {
         BRunUtil.invoke(result, "testReadOnlyListWithListBindingPatternInVarDecl");
     }
 
+    @Test
+    public void testTupleVariableWithAnonymousRecordType() {
+        BRunUtil.invoke(result, "testTupleVariableWithAnonymousRecordType");
+    }
+
     private void validateTupleVarDefWithUnitionComplexResults(BArray returns) {
         Assert.assertEquals(returns.size(), 3);
 

--- a/tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal
+++ b/tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal
@@ -744,15 +744,11 @@ function testReadOnlyRecordWithMappingBindingPatternInVarDecl() {
 }
 
 function recordVariableWithAnonymousType() returns [string, boolean] {
-    do {
-        record {|
-            string name;
-            boolean married;
-        |} {name, married} = check getPersonRecord().ensureType();
-        return [name, married];
-    } on fail {
-        return ["name", false];
-    }
+    record {|
+        string name;
+        boolean married;
+    |} {name, married} = checkpanic getPersonRecord().ensureType();
+    return [name, married];
 }
 
 function getPersonDetails() returns PersonDetails {
@@ -767,16 +763,12 @@ type PersonDetails record {|
 |};
 
 function nestedRecordVariableWithAnonymousType() returns [string, boolean, record {|int age;|}] {
-    do {
-        record {|
-            string name;
-            boolean married;
-            record {|int age;|} details;
-        |} {name, married, details} = check getPersonDetails().ensureType();
-        return [name, married, details];
-    } on fail {
-        return ["name", false, {age: 0}];
-    }
+    record {|
+        string name;
+        boolean married;
+        record {|int age;|} details;
+    |} {name, married, details: {age}} = checkpanic getPersonDetails().ensureType();
+    return [name, married, {age}];
 }
 
 function testRecordVariableWithAnonymousType() {

--- a/tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal
+++ b/tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal
@@ -743,12 +743,13 @@ function testReadOnlyRecordWithMappingBindingPatternInVarDecl() {
     assertEquality(<ReadOnlyRecord> {x: [1, 2], y: "s1"}, b1);
 }
 
-function recordVariableWithAnonymousType() returns [string, boolean] {
+function testRecordVariableWithAnonymousType() {
     record {|
         string name;
         boolean married;
     |} {name, married} = checkpanic getPersonRecord().ensureType();
-    return [name, married];
+    assertEquality("Jack", name);
+    assertTrue(married);
 }
 
 function getPersonDetails() returns PersonDetails {
@@ -762,25 +763,14 @@ type PersonDetails record {|
     record {|int age;|} details;
 |};
 
-function nestedRecordVariableWithAnonymousType() returns [string, boolean, record {|int age;|}] {
+function testNestedRecordVariableWithAnonymousType() {
     record {|
         string name;
         boolean married;
         record {|int age;|} details;
     |} {name, married, details: {age}} = checkpanic getPersonDetails().ensureType();
-    return [name, married, {age}];
-}
-
-function testRecordVariableWithAnonymousType() {
-    [string, boolean] [name, married] = recordVariableWithAnonymousType();
     assertEquality("Jack", name);
-    assertTrue(married);
-}
-
-function testNestedRecordVariableWithAnonymousType() {
-    [string, boolean, record {|int age;|}] [name, married, details] = nestedRecordVariableWithAnonymousType();
-    assertEquality("Jack", name);
-    assertEquality(30, details.age);
+    assertEquality(30, age);
     assertTrue(married);
 }
 

--- a/tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal
+++ b/tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal
@@ -743,17 +743,52 @@ function testReadOnlyRecordWithMappingBindingPatternInVarDecl() {
     assertEquality(<ReadOnlyRecord> {x: [1, 2], y: "s1"}, b1);
 }
 
-function recordVariableWithAnonymousType() returns [string, boolean]|error {
-    record {|
-        string name;
-        boolean married;
-    |} {name, married} = check getPersonRecord().ensureType();
-    return [name, married];
+function recordVariableWithAnonymousType() returns [string, boolean] {
+    do {
+        record {|
+            string name;
+            boolean married;
+        |} {name, married} = check getPersonRecord().ensureType();
+        return [name, married];
+    } on fail {
+        return ["name", false];
+    }
+}
+
+function getPersonDetails() returns PersonDetails {
+    PersonDetails details = {name: "Jack", married: true, details: {age: 30}};
+    return details;
+}
+
+type PersonDetails record {|
+    string name;
+    boolean married;
+    record {|int age;|} details;
+|};
+
+function nestedRecordVariableWithAnonymousType() returns [string, boolean, record {|int age;|}] {
+    do {
+        record {|
+            string name;
+            boolean married;
+            record {|int age;|} details;
+        |} {name, married, details} = check getPersonDetails().ensureType();
+        return [name, married, details];
+    } on fail {
+        return ["name", false, {age: 0}];
+    }
 }
 
 function testRecordVariableWithAnonymousType() {
-    [string, boolean]|error [name, married] = recordVariableWithAnonymousType();
+    [string, boolean] [name, married] = recordVariableWithAnonymousType();
     assertEquality("Jack", name);
+    assertTrue(married);
+}
+
+function testNestedRecordVariableWithAnonymousType() {
+    [string, boolean, record {|int age;|}] [name, married, details] = nestedRecordVariableWithAnonymousType();
+    assertEquality("Jack", name);
+    assertEquality(30, details.age);
     assertTrue(married);
 }
 

--- a/tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal
+++ b/tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal
@@ -743,6 +743,20 @@ function testReadOnlyRecordWithMappingBindingPatternInVarDecl() {
     assertEquality(<ReadOnlyRecord> {x: [1, 2], y: "s1"}, b1);
 }
 
+function recordVariableWithAnonymousType() returns [string, boolean]|error {
+    record {|
+        string name;
+        boolean married;
+    |} {name, married} = check getPersonRecord().ensureType();
+    return [name, married];
+}
+
+function testRecordVariableWithAnonymousType() {
+    [string, boolean]|error [name, married] = recordVariableWithAnonymousType();
+    assertEquality("Jack", name);
+    assertTrue(married);
+}
+
 //////////////////////////////////////////////////////////////////////////////////////
 
 const ASSERTION_ERROR_REASON = "AssertionError";

--- a/tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/tuple-variable-definition.bal
+++ b/tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/tuple-variable-definition.bal
@@ -300,16 +300,11 @@ type PersonDetails [record {|
     boolean married;
 |}, int];
 
-function tupleVariableWithAnonymousRecordType() returns [string, boolean, int] {
+function testTupleVariableWithAnonymousRecordType() {
     [record {|
         string name;
         boolean married;
     |}, int] [{name, married}, age] = checkpanic getPersonDetails().ensureType();
-    return [name, married, age];
-}
-
-function testTupleVariableWithAnonymousRecordType() {
-    [string, boolean, int] [name, married, age] = tupleVariableWithAnonymousRecordType();
     assertEquality("Jack", name);
     assertEquality(true, married);
     assertEquality(10, age);

--- a/tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/tuple-variable-definition.bal
+++ b/tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/tuple-variable-definition.bal
@@ -290,6 +290,31 @@ function testReadOnlyListWithListBindingPatternInVarDecl() {
     assertEquality(<ReadOnlyTuple> [[1, 2], "s1"], b1);
 }
 
+function getPersonDetails() returns PersonDetails {
+    PersonDetails details = [{name: "Jack", married: true}, 10];
+    return details;
+}
+
+type PersonDetails [record {|
+    string name;
+    boolean married;
+|}, int];
+
+function tupleVariableWithAnonymousRecordType() returns [string, boolean, int] {
+    [record {|
+        string name;
+        boolean married;
+    |}, int] [{name, married}, age] = checkpanic getPersonDetails().ensureType();
+    return [name, married, age];
+}
+
+function testTupleVariableWithAnonymousRecordType() {
+    [string, boolean, int] [name, married, age] = tupleVariableWithAnonymousRecordType();
+    assertEquality("Jack", name);
+    assertEquality(true, married);
+    assertEquality(10, age);
+}
+
 const ASSERTION_ERROR_REASON = "AssertionError";
 
 function assertEquality(any expected, any actual) {

